### DISCUSSION TITLE
Is there a way to know if a device is connected via AirPlay?

### DISCUSSION THREAD

**Discussion by UngluedChalice** [mikebrady/shairport-sync#1397] [Ideas]
I currently use Home Assistant and Node-Red to start a timer when I turn on my HTD whole-house audio system and alert me if the system has been on for longer than a given amount of time. I have my RaspberryPi running Shairport-sync hooked up to the whole-house audio system, and then AirPlay to the pi to get the audio from my iPhone to the whole-house audio system. 

Sometimes people forget to turn the system off, and it just gets left on overnight. I want to avoid that, hence the timer. 

A better method to make sure it doesn't get left on would be for Home Assistant/Node-Red to be able to check and see if there is a device connected to Shairport-Sync via AirPlay. If there is no device connected and the system is on, it can trigger an alert to let me know. But to do this, I need for Home Assistant/Node-Red to be able to check and see if someone is connected. 

Anyone know if this is possible?

**Comment by noelhibbard** [comment]
I would look at the sessioncontrol section of shairport-sync.conf. You will see a bunch of events that can trigger scripts. You could use these events to accomplish what you want to do. One of my shairport-sync instances uses these events to turn on an AVR and then when playback ends I call a script that starts a 10 minute timer and if the timer isn't interrupted by another event before it times out, it turns off the AVR.

You could also use shairport-sync's MQTT integration to capture playback start/stop events.

The tools are there, you just have to get creative.

**Reply by UngluedChalice** [reply]
Okay cool, thanks for replying. 

So I found the relevant section:

```
// Advanced parameters for controlling how Shairport Sync stays active and how it runs a session
sessioncontrol =
{
//	"active" state starts when play begins and ends when the active_state_timeout has elapsed after play ends, unless another play session starts before the timeout has fully elapsed.
//	run_this_before_entering_active_state = "/full/path/to/application and args"; // make sure the application has executable permission. If it's a script, include the shebang (#!/bin/...) on the first line
//	run_this_after_exiting_active_state = "/full/path/to/application and args"; // make sure the application has executable permission. If it's a script, include the shebang (#!/bin/...) on the first line
//	active_state_timeout = 10.0; // wait for this number of seconds after play ends before leaving the active state, unless another play session begins.

//	run_this_before_play_begins = "/full/path/to/application and args"; // make sure the application has executable permission. If it's a script, include the shebang (#!/bin/...) on the first line
//	run_this_after_play_ends = "/full/path/to/application and args"; // make sure the application has executable permission. If it's a script, include the shebang (#!/bin/...) on the first line

//	run_this_if_an_unfixable_error_is_detected = "/full/path/to/application and args"; // if a problem occurs that can't be cleared by Shairport Sync itself, hook a program on here to deal with it. An error code-string is passed as the last argument.
//	  Many of these "unfixable" problems are caused by malfunctioning output devices, and sometimes it is necessary to restart the whole device to clear the problem.
//	  You could hook on a program to do this automatically, but beware -- the device may then power off and restart without warning!
//	wait_for_completion = "no"; // set to "yes" to get Shairport Sync to wait until the "run_this..." applications have terminated before continuing

//	allow_session_interruption = "no"; // set to "yes" to allow another device to interrupt Shairport Sync while it's playing from an existing audio source
//	session_timeout = 120; // wait for this number of seconds after a source disappears before terminating the session and becoming available again.
};
```

My idea would be to have something run using this: `run_this_after_exiting_active_state = "/full/path/to/application and args";`

However, I have no idea what that something is, as I'm super new to all of this. My current setup is a raspberry pi running shairport-sync and a few other things. I have a second pi dedicated to Home Assistant. I would like to somehow let the second pi running Home Assistant know when shairport-sync exits the active state. 

Is this something a script could do? Any pointers on how to do this or what to google to learn more? I've heard of MQTT, but never used it. If I could get shairport-sync to flip an input_boolean on Home Assistant or something, that would be great. I just have no idea how to get one pi to send a message to the other - I'm a bit over my head! If you could even just post what you use, it would give me an example to look at and maybe learn from.

**Reply by noelhibbard** [reply]
I don't have any experience with HA but I suspect it has a REST API. If so, you could use curl to hit the REST API from your shairport-sync Pi. The only problem is ```run_this_after_exiting_active_state``` is called as soon as playback stops... for example a pause would also trigger it. So you need to start a timer when playback stops and cancel the timer when playback starts. When the timer runs out, you would call the HA REST API. You could set the timer for 10mins so it wouldn't mark shairport-sync as inactive until it's been stopped for 10mins. I guess you could do it via scripts.

The stop script could be named ```shairport-sync-stop.sh``` and look something like this:
```bash
#!/usr/bin/env bash

sleep 10m
curl http://hapi.example.com/some/rest/endpoint/to/mark/sps/active
```

And the start script would be named ```shairport-sync-start.sh``` and look something like this:
```bash
#!/usr/bin/env bash

pkill -f shairport-sync-stop.sh
curl http://hapi.example.com/some/rest/endpoint/to/mark/sps/inactive
```

You can see that the stop script sleeps for 10 minutes and then calls a URL telling HA that SPS is no longer in use. The start script attempts to kill the stop script if it's running and then calls a URL telling HA that SPS is in use. Sorry I don't have more knowledge on HA to be able to tell you exact URLs to accomplish what you want.

**Comment by stephenmetzger** [comment]
If you have HA running, it should be super easy to get an mqtt (mosquito) host running - there is an add-on in the HA Addon Store. 

For a new user, this should be mush easier than using scripts.

**Reply by UngluedChalice** [reply]
Yes, I have MQTT on HA, but I've never used it. Assuming I can figure out the HA end, what would I need to do on the shair-port sync end to get it to send "an MQTT message" to HA?

**Reply by stephenmetzger** [reply]
First, set up the Mosquitto broker add-on in the Add-on Store in HA. Then, add the MQTT integration to HA. 

You may need to reinstall Shairport on your pi using the --with-mqtt-client flag. 

Then, in your shairport-sync.conf, uncomment and edit the relevant lines to configure mqtt:

```// How to enable the MQTT-metadata/remote-service
// For this section to be operative, Shairport Sync must be built with the 
following configuration flag:
// --with-mqtt-client
mqtt =
{
        enabled = "yes"; // set this to yes to enable the mqtt-metadata-service
        hostname = "[MQTT Broker IP ADDRESS]"; // Hostname of the MQTT 
Broker
        port = 1883; // Port on the MQTT Broker to connect to
        username = "[MQTT USERNAME - IF SET]"; //set this to a string to your 
username in order to enable username authentication
        password = "[MQTT PASSWORD - IF SET]"; //set this to a string you your 
password in order to enable username & password authentication
//      capath = NULL; //set this to the folder with the CA-Certificates to be 
accepted for the server certificate. If not set, TL$
//      cafile = NULL; //this may be used as an (exclusive) alternative to capath 
with a single file for all ca-certificates
//      certfile = NULL; //set this to a string to a user certificate to enable MQTT 
Client certificates. keyfile must also be set!
//      keyfile = NULL; //private key for MQTT Client authentication
        topic = "[NAME OF YOUR SHAIRPORT INSTANCE - CAN BE ANYTHING]"; 
//MQTT topic where this instance of shairport-sync should publish. If not set, 
the general.name$
        publish_raw = "no"; //whether to publish all available metadata under 
the codes given in the 'metadata' docs.
        publish_parsed = "yes"; //whether to publish a small (but useful) subset 
of metadata under human-understandable topics
//      Currently published 
topics:artist,album,title,genre,format,songalbum,volume,client_ip,
//      Additionally, empty messages at the topics  
play_start,play_end,play_flush,play_resume are published
        publish_cover = "yes"; //whether to publish the cover over mqtt in binary 
form. This may lead to a bit of load on the brok$
        enable_remote = "yes"; //whether to remote control via MQTT. RC is 
available under `topic`/remote.
//      Available commands are "command", "beginff", "beginrew", 
"mutetoggle", "nextitem", "previtem", "pause", "playpause", "play$
};
```

Restart shairport-sync and use ssh, HA, or tool like MQTTfx on Mac/PC or an MQTT smartphone app to connect to your mqtt broker and check to make sure messages are being sent. In a tool like MQTTfx, you'd connect to the broker and then subscribe to the topic (whatever you used in the "topic" field of the config file. For example, I have a SPS instance set up with the topic "officeshairport", so to subscribe to all messages on "officeshairport/#". 

As you start, stop, pause, skip, ect. you should see messages under topics like "officeshairport/play_start", "officeshairport/play_end", etc. 

If you wanted to use these mqtt commands to toggle an input boolean in HA, you could set up an automation that uses the MQTT platform as a trigger. Something like this would turn the input boolean to on when a playback session on SPS started:

```alias: New Automation
description: ''
mode: single
trigger:
  - platform: mqtt
    topic: officeshairport/play_start
condition: []
action:
  - service: input_boolean.turn_on
    target:
      entity_id: input_boolean.massive_speaker_system
```
      
You could also set up MQTT as a trigger in nodred very easilty. Sorry if this is all confusing. Its a lot of little steps. I'm happy to help walk you through it. Easier to message me on Discord?

**Reply by UngluedChalice** [reply]
Thanks for writing this out! Haven't been able to sit down and work through it until now, but I was able to get it all working. Even figured out how to install `libmosquitto-dev` as that was required. These are big accomplishments for me, as I'm basically just a monkey copying and pasting!

For anyone else that comes along and finds this, here is what that portion of the `.conf` file looks like:

```
mqtt =
{
        enabled = "yes";
        hostname = "192.168.1.40";
        port = 1883;
        username = "mqtt-user";
        password = "PASSWORD FOR mqtt-user in Home Assisstant";
        topic = "AirPlay";
        publish_raw = "no";
        publish_parsed = "yes";
```

I see the MQTT messages in Node Red and HA, so I should be able to look for the `AirPlay/play_end` message and then automatically turn off the speaker system after a set amount of time unless it sees the `AirPlay/play_start` message. 

Any ideas on how to automate the process of playing music further? Currently I have to turn on whatever speaker zones I want in HA. Then connect via AirPlay. Then play the music. I was hoping for Shairport-sync to spit out a message when I initially connect to it via AirPlay, but that doesn't seem to be happening. Then HA can push out a notification that can have some actions on it so I could quickly turn on my most frequently used zones or be taken to the Speakers dashboard in Lovelace.

I could go really crazy and do some sort of Bluetooth presence detection I've heard about and then automatically turn the speakers on as I walk around the house! Ha!

Edit to Add: the most cumbersome part of the process is the connecting to the AirPlay part. I wish I could somehow use HA to trigger an iOS automation or something that would then automatically connect my iPhone to the Shairport-sync AirPlay server.

**Reply by stephenmetzger** [reply]
Edit: Replace long code blocks w/ pastebin links. 

No problem. If I knew more about how your multi-room speaker system is set up I could for sure help you out with some ideas. Sounds like you have a multi-zone amp that integrates with HA and exposes different switch entities for each zone?

The nodered flow that I use to power on and off my speakers depending on SPS status is:

https://pastebin.com/VntW73Wk

Also, check out the Shairport Sync Media Player integration in HACS (https://github.com/parautenbach/hass-shairport-sync). This will expose your shairport instances as media_player entities in HA. If you have meta-data including album artwork set up in SPS can set up a nice media player card in Lovelace (I recommend using mini-media-player, also in HACS - https://github.com/kalkih/mini-media-player):

https://imgur.com/a/w3Z7rrI

I have 3 SPS instances running, so I use the HA conditional card to only show the media player cards if the state of the media_player is "playing". 

I'm still playing around with the SPS Media Player integration to try and get it to do some more advanced stuff (like change the volume level of the pi running SPS on media player card volume up and down button presses - the volumeup and volumedown mqtt commands for SPS change the volume of all instances in an active Airplay2 'session' - not just the instance you send the command to, which is a known limitation right now. So, if I want to have control from HA over the volume of each individual instance while using AirPlay2, I have to make the change to the pi mixer. 

Also, the Power Button on the SPS HACS integration doesn't do anything, so I hijacked it to perform a reset of SPS, which helps if something freezes up, and also acts to disconnect that SPS instance from a current AP2 session (but doesn't interrupt the session on the other SPS instances). So, my plan is to get something like an IKEA Tradfri controller (https://www.ikea.com/us/en/p/tradfri-remote-control-00443130/) and map the buttons to next track, previous track, volume up, volume down, play/pause, and disconnect - and put one in each room with an SPS instance. The flow I use to send the shairport restart command is here:

https://pastebin.com/iphrkNsb

Edit: The active_start message should be sent when you start playing your media, not on connect to the SPS instance.  

Explain what you mean by "connecting to the airplay part". If you are going to be casting music to AirPlay, wont you aleady have to have your phone open so that you can select what music to play? 

Apple does have an option in "Shortcuts" that will automatically connect you to a known ApirPlay source, and I think it can even start playing a specific playlist, station, etc. But, it would still need a trigger - could be a homescreen shortcut, registering an NFC tag, etc. Also, if you have homebridge set up in HA you should can get HA entities in the Home App and use Home App Automations to trigger shortcuts. Lots of options.

**Reply by stephenmetzger** [reply]
Oh, also, your SPS instance should be discoverable in the iOS Home App. You'll have to go into the Home App, select "+", then "Add Accessory", and you might have to top "I Don't Have A Code of Cannot Scan", which should bring you to a screen where compatible HomeKit devices are discovered. If all is well, your SPS instance will be discoverd. Once set up, you should be able to use Siri to start casting to AirPlay. You may have to play around with it, I usually invoke it with the name of the room I want to stream to - my bathroom SPS is in homekit - named Bathroom Speakers and  in the room "Bathroom". So if I say, "Hey Siri, play the Dance Music by The Mountain Goats in the bathroom", SPS will start a stream to the bathroom, which will power on my speakers, and music will start playing.

**Reply by UngluedChalice** [reply]
So I have an [HTD Lync 12 system](https://www.htd.com/Products/Whole-House-Audio/Lync) for whole house audio via hardwired speakers in the ceiling of each room. This system can have like 12 different sources, but I just have a single source: the pi running SPS. So I AirPlay to the pi from my phone and then the whole house audio plays that over the speakers. 

[This is the integration I'm using](https://github.com/charger68/htd-lync12-home-assistant) to pull the HTD stuff into Home Assistant. HTD talked about releasing an API and all this stuff, but never has moved with it. [Here is the thread](https://community.home-assistant.io/t/integrating-htd-into-homeassistant/321797/15) of me getting help trying to get it to work, for reference in case someone comes along later and needs that info for themselves. 

I looked at your node read flow. Why do you have the topic as `/active_start` instead of `/play_start`? I searched on this github repo but only found the base code.

At this point I don't think I'll put the media player in HA. I guess I don't see the point for me, as I'm almost always playing stuff with my phone, so all the controls are on the lock screen. Is there some other benefit I'm missing? And I'm not needing AirPlay2, as I have the whole-house audio for that. I did look into AirPlay2, but at that point no one had cracked it yet, so I went with this system. 

I like the idea of a button to start and stop SPS. Do you use this command? `sudo systemctl start shairport-sync` and `sudo systemctl stop shairport-sync` or is there something better?

> Explain what you mean by "connecting to the airplay part". If you are going to be casting music to AirPlay, wont you aleady have to have your phone open so that you can select what music to play?

So from my phone I pull down the control center, then tap the airplay button, then maybe wait a second for my SPS instance to appear, then tap it, then tap to close that panel, then tap again to close the control center. It would be awesome if there was a way to automatically AirPlay to SPS from my phone whenever I turned on a speaker. 

I'm going to have to look into the Shortcuts - this could be really helpful. At work (I'm a teacher) we project via AirPlay, so being have to tap a button to connect to the correct room would be nice, as when I go to project there is a long list of AirPlay devices I could connect to. 

So I don’t have HomeKit (I think that's what its called) running at all. I have the HomeKit Controller running on Home Assistant, so I can then control my ecobee that way. But I don't have an AppleTV or anything running HomeKit. Do I need that for HomeBridge? 

Edit: any way to trigger an iOS shortcut from Home Assistant? That could make it so when I turn on my speakers, my phone automatically AirPlays to SPS. Hmm...

With respect to finding the SPS instance in the iOS Home app, when I get to the "Add Accessory" page I get my Lutron Smart Bridge Pro 2 to appear, at times (but not consistently) my Yamaha receiver which has AirPlay, but never the SPS instance. 

_I am gonna make it through this year, if it kills me..._

**Comment by stephenmetzger** [comment]
You could use play_start and play_end if you want to. Download MQTTfx and play around, watching the messages come in. 

I didn't want the speaker power to toggle on and off every time I paused and then resumed play, so I went with active_start and active_end. The "play" messages get sent on any play/pause action. The active_start message is sent when a "session" starts (i.e. you connect to the source and start streaming to it - you dont get the message just by connecting, you have to also start streaming). Active end is sent when disconnected from the source, after a delay as specific in the SPS config file (see active_state_timeout). 

I like the media player in HA just for aesthetics. Cool to have a card that shows metadata and album art, especially on a wall-mounted tablet.

SPS has a beta version that has been super stable for me and implements AP2, it rules, but yeah - doesn't matter if you already use a multichannel relay for whole house. 

Use `sudo systemctl restart shairport-sync`. 

That is a good question. I may be wrong, but i think you can still run homebridge, and have use iOS Home locally with no AppleTV/iPad acting as a Home Hub. But unsure. 

Yes, siri shortcuts can be triggered by HA:

https://companion.home-assistant.io/docs/integrations/siri-shortcuts/

As for not seeing SPS in iOS Home, what version of SPS are you running (run `shairport-sync -V` on the pi running SPS). I'm running 4.0-dev-312-g687fd167, and working well. Maybe the main branch doesn't have support for HomeKit integration?

Example of a shortcut that connects to a specific AirPlay Source, opens the Apple Music app, and starts playing is here, you should be able to play around and tailor it to your needs:

https://www.icloud.com/shortcuts/99fb471146f342daab8992ebb09f9441

**Reply by UngluedChalice** [reply]
Ah, so what I did was set up a little timer that gets stopped if I start playing again. [Here's the node-red sequence.](https://gist.github.com/UngluedChalice/0f2b535252d55d88c7bfef6b09055529)
[Here is a screenshot](https://imgur.com/a/h9pJyS3) of what it looks like in node-red. I got the structure of that flow from [Mostly Chris on YouTube](https://www.youtube.com/watch?v=PSybOWztoS0), so gotta give credit where credit is due. 

Yeah I was looking at that shortcut link earlier - I think the only way to do it is via a notification. I'll have to look into that more. Maybe something where I turn on the speakers I get the notification that I can tap and trigger the shortcut to connect and open Apple Music? Apple isn't going to let something run in the background, which would be ideal. 

I am not on the dev channel, I'm on `3.3.9-OpenSSL-Avahi-ALSA-metadata-mqtt-sysconfdir:/usr/local/etc` That would be up-to-date as of today.

**Reply by stephenmetzger** [reply]
Hmm. Not sure which versions support HomeKit. @mikebrady, does 3.3.9 support homekit?

**Reply by charlesomer** [reply]
HomeKit is currently only supported on the development branch

**Reply by noelhibbard** [reply]
4.0+ on the dev branch is needed for AirPlay2/HomeKit.

**Comment by lenkunz** [comment]
As I dig through this. If the protocol is airplay 2 and the raw payload is configured.
When device connected, it will be publish the client's IP through the topic [topic]/ssnc/conn
And when disconnected, it will be publish the IP through to the [topic]/ssnc/disc

**Reply by Vaskyy** [reply]
Im trying to setup an home assistant sensor for this.
I have a couple of players that are being powered on and off.
But i do not get anything at the configured mqtt topics.

When its available i want home assistant to know this, so music assistant can include the player in its sync_group.
And when its unpowered and the timeout is sent through mqtt, this could trigger an leave sync_group command.

Can someone point me in the right direction, in how to setup something like that?

I do read connect and disconnect in mqtt broker log, but cant figure out how to make sensor out of that:
```
2025-05-17 16:04:39: New client connected from 192.168.178.37:50880 as Vaskyy-Box.
2025-05-17 16:08:11: Client Vaskyy-Box has exceeded timeout, disconnecting.
```

**Reply by Vaskyy** [reply]
Not when device is being connected and disconnected.

I'd need a "AirPlay is ready to connect to"-sensor, so the playing device can turn on playing to that specific speaker, without my need to change that manually.

I've got a kitchen speaker which starts with my presence.
Rn its:
1. Presence
2. Smartplug turns on RPI with shairport-sync installed
3. 1:10 minute wait (until PI and shairport-sync is launched)
4. send music to that speaker manually (or tru Music Assistant join speaker feature)

So without the smartplug it wouldn't be possible to track when the speaker is connectable.

With an "Hey im here"-MQTT message from that shairport-instance, it would be perfect to listen to that, and THEN start the streaming, instead of waiting a set amount of time, after turning the whole device on.

**Reply by mikebrady** [reply]
Thanks for the questions! We are really at the mercy of whatever information can be obtained from the AirPlay system and then a little from Shairport Sync itself. All that information is provided in the basic metadata stream. The MQTT system is a user of that metadata. So, the best way to find out what information is available is to monitor the metadata stream, possibly with the [sample metadata reader](https://github.com/mikebrady/shairport-sync-metadata-reader). If there is something in there that is useful and is not made accessible on MQTT, then let us know. Note that the information you can get from AirPlay and AirPlay 2 is different.
